first, I need to import the necessart libraries for now; I will be adding more later on as I need them. 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

now, I need to import my dataframe into this file.

In [2]:
my_data = pd.read_csv("IBM_people_dataset.csv")

In [4]:
pd.set_option('display.max_columns', None) #because I need to see the whole data

In [5]:
my_data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [6]:
my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
Age                         1470 non-null int64
Attrition                   1470 non-null object
BusinessTravel              1470 non-null object
DailyRate                   1470 non-null int64
Department                  1470 non-null object
DistanceFromHome            1470 non-null int64
Education                   1470 non-null int64
EducationField              1470 non-null object
EmployeeCount               1470 non-null int64
EmployeeNumber              1470 non-null int64
EnvironmentSatisfaction     1470 non-null int64
Gender                      1470 non-null object
HourlyRate                  1470 non-null int64
JobInvolvement              1470 non-null int64
JobLevel                    1470 non-null int64
JobRole                     1470 non-null object
JobSatisfaction             1470 non-null int64
MaritalStatus               1470 non-null object
MonthlyIncome         

We do not have any mssing data, since all of the categories have 1470 observations. However, I can see that some of the data are categorical (object), so I need to turn them into numerical. I can also see that there are some variables that I do not need obviously in my algorithm, so I will try to understand better which variables are useless and remove them from the dataframe. First let me see what are the categorical variables composed of so that I can understand and analyze them further. 

In [7]:
my_data.Attrition.unique()

array(['Yes', 'No'], dtype=object)

In [8]:
my_data.BusinessTravel.unique() #can be turned into an ordinal numerical variable

array(['Travel_Rarely', 'Travel_Frequently', 'Non-Travel'], dtype=object)

In [9]:
my_data.Department.unique()

array(['Sales', 'Research & Development', 'Human Resources'], dtype=object)

In [10]:
my_data.EducationField.unique()

array(['Life Sciences', 'Other', 'Medical', 'Marketing',
       'Technical Degree', 'Human Resources'], dtype=object)

In [11]:
my_data.Gender.unique()

array(['Female', 'Male'], dtype=object)

In [12]:
my_data.JobRole.unique()

array(['Sales Executive', 'Research Scientist', 'Laboratory Technician',
       'Manufacturing Director', 'Healthcare Representative', 'Manager',
       'Sales Representative', 'Research Director', 'Human Resources'],
      dtype=object)

In [13]:
my_data.MaritalStatus.unique()

array(['Single', 'Married', 'Divorced'], dtype=object)

In [14]:
my_data.Over18.unique()

array(['Y'], dtype=object)

In [15]:
my_data.OverTime.unique()

array(['Yes', 'No'], dtype=object)

In [16]:
my_data = my_data.drop("Over18", axis=1) #since it has the same value for everyone

In [17]:
my_data = my_data.drop("JobRole", axis=1) #since it is extremely similar to and shows the same thing as "Department"

In [18]:
my_data = my_data.drop("EmployeeNumber", axis=1) #since it is unique for every obersvation and is useless since it shows a random number

In [19]:
my_data.EmployeeCount.unique()

array([1], dtype=int64)

In [20]:
my_data = my_data.drop("EmployeeCount", axis=1) #since again it only has the one same value for every observation and is completely useless

In [22]:
my_data.Education.unique() #not dropping this

array([2, 1, 4, 3, 5], dtype=int64)

In [23]:
my_data.StandardHours.unique() #of course dropping this for same reasons mentioned above

array([80], dtype=int64)

In [24]:
my_data = my_data.drop("StandardHours", axis=1)

In [28]:
my_data.EnvironmentSatisfaction.unique() #not dropping this either, just observing what it is

array([2, 3, 4, 1], dtype=int64)

now I will turn all the categorical variables into numerical

In [29]:
my_data.info()  #just remembering what I have left

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 30 columns):
Age                         1470 non-null int64
Attrition                   1470 non-null object
BusinessTravel              1470 non-null object
DailyRate                   1470 non-null int64
Department                  1470 non-null object
DistanceFromHome            1470 non-null int64
Education                   1470 non-null int64
EducationField              1470 non-null object
EnvironmentSatisfaction     1470 non-null int64
Gender                      1470 non-null object
HourlyRate                  1470 non-null int64
JobInvolvement              1470 non-null int64
JobLevel                    1470 non-null int64
JobSatisfaction             1470 non-null int64
MaritalStatus               1470 non-null object
MonthlyIncome               1470 non-null int64
MonthlyRate                 1470 non-null int64
NumCompaniesWorked          1470 non-null int64
OverTime               

In [30]:
Churn = pd.get_dummies(my_data.Attrition)

In [31]:
my_data = my_data.drop("Attrition", axis=1)

In [32]:
Churn = Churn.drop("No", axis=1)  #now 1s are churners, 0s are non-churners

In [33]:
my_data = my_data.join(Churn) #called Yes, will rename later

In [34]:
my_data.BusinessTravel = my_data.BusinessTravel.astype("category").cat.reorder_categories(["Non-Travel","Travel_Rarely","Travel_Frequently"]).cat.codes

In [35]:
Department = pd.get_dummies(my_data.Department)

In [36]:
my_data = my_data.drop("Department", axis=1)

In [37]:
Department = Department.drop("Human Resources", axis=1) #due to DoF

In [38]:
my_data=my_data.join(Department)

In [39]:
ed_field = pd.get_dummies(my_data.EducationField)

In [40]:
my_data = my_data.drop("EducationField", axis=1)

In [41]:
ed_field =ed_field.drop("Other", axis=1)

In [42]:
my_data = my_data.join(ed_field)

In [43]:
gender = pd.get_dummies(my_data.Gender)

In [44]:
my_data = my_data.drop("Gender", axis=1)

In [45]:
gender = gender.drop("Male", axis=1) #now 1s are female, 0s are male

In [46]:
my_data = my_data.join(gender) #now it's called "Female" but i'll change the name later

In [47]:
marital_status = pd.get_dummies(my_data.MaritalStatus)

In [48]:
my_data = my_data.drop("MaritalStatus", axis=1)

In [49]:
marital_status = marital_status.drop("Divorced", axis=1)

In [50]:
my_data = my_data.join(marital_status)

In [51]:
overtime = pd.get_dummies(my_data.OverTime)

In [52]:
my_data = my_data.drop("OverTime", axis=1)

In [53]:
ovrtime = overtime.drop("No", axis=1) #now 1s are overtime, 0s are not

before adding overtime to my data, i'll change the names so that the "Yes" of overtime doesn't overlap with the "yes" of churn

In [54]:
my_data=my_data.rename(columns = {"Female" : "Gender", "Yes" : "Churn", })

In [55]:
overtime = overtime.rename(columns = {"Yes" : "Overtime"})

In [56]:
my_data = my_data.join(overtime)

now let me make sure everything is ok with my data

In [57]:
my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 37 columns):
Age                         1470 non-null int64
BusinessTravel              1470 non-null int8
DailyRate                   1470 non-null int64
DistanceFromHome            1470 non-null int64
Education                   1470 non-null int64
EnvironmentSatisfaction     1470 non-null int64
HourlyRate                  1470 non-null int64
JobInvolvement              1470 non-null int64
JobLevel                    1470 non-null int64
JobSatisfaction             1470 non-null int64
MonthlyIncome               1470 non-null int64
MonthlyRate                 1470 non-null int64
NumCompaniesWorked          1470 non-null int64
PercentSalaryHike           1470 non-null int64
PerformanceRating           1470 non-null int64
RelationshipSatisfaction    1470 non-null int64
StockOptionLevel            1470 non-null int64
TotalWorkingYears           1470 non-null int64
TrainingTimesLastYear       14

I have no idea what No is

In [58]:
my_data.head()

Age  BusinessTravel  DailyRate  DistanceFromHome  Education  \
0   41               1       1102                 1          2   
1   49               2        279                 8          1   
2   37               1       1373                 2          2   
3   33               2       1392                 3          4   
4   27               1        591                 2          1   

   EnvironmentSatisfaction  HourlyRate  JobInvolvement  JobLevel  \
0                        2          94               3         2   
1                        3          61               2         2   
2                        4          92               2         1   
3                        4          56               3         1   
4                        1          40               3         1   

   JobSatisfaction  MonthlyIncome  MonthlyRate  NumCompaniesWorked  \
0                4           5993        19479                   8   
1                2           5130        24907                   1   
2                3           2090         2396                   6   
3                3           2909        23159                   1   
4                2           3468        16632                   9   

   PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
0                 11                  3                         1   
1                 23                  4                         4   
2                 15                  3                         2   
3                 11                  3                         3   
4                 12                  3                         4   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  Churn  \
0                        0                     5      1   
1                        1                     7      0   
2                        0                     0      1   
3                        3                     0      0   
4                        2                     2      0   

   Research & Development  Sales  Human Resources  Life Sciences  Marketing  \
0                       0      1                0              1          0   
1                       1      0                0              1          0   
2                       1      0                0              0          0   
3                       1      0                0              1          0   
4                       1      0                0              0          0   

   Medical  Technical Degree  Gender  Married  Single  No  Overtime  
0        0                 0       1        0       1   0         1  
1        0                 0       0        1       0   1         0  
2        0                 0       0        0       1   0         1  
3        0                 0       1        1       0   0         1  
4        1                 0       0        1       0   1         0

ok so No is the opposite of Overtime, I guess it wasn't dropped. No problem.

In [59]:
my_data = my_data.drop("No", axis=1)

In [60]:
my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 36 columns):
Age                         1470 non-null int64
BusinessTravel              1470 non-null int8
DailyRate                   1470 non-null int64
DistanceFromHome            1470 non-null int64
Education                   1470 non-null int64
EnvironmentSatisfaction     1470 non-null int64
HourlyRate                  1470 non-null int64
JobInvolvement              1470 non-null int64
JobLevel                    1470 non-null int64
JobSatisfaction             1470 non-null int64
MonthlyIncome               1470 non-null int64
MonthlyRate                 1470 non-null int64
NumCompaniesWorked          1470 non-null int64
PercentSalaryHike           1470 non-null int64
PerformanceRating           1470 non-null int64
RelationshipSatisfaction    1470 non-null int64
StockOptionLevel            1470 non-null int64
TotalWorkingYears           1470 non-null int64
TrainingTimesLastYear       14

ok everything seems fine except for the fact that there are a lot of variables in the dataframe, so I will try to understand which variables are highly correlated so that I can remove them from the model

In [62]:
correlation_1 = my_data.corr(method = "pearson")

In [63]:
correlation_1

Age  BusinessTravel  DailyRate  \
Age                       1.000000       -0.011807   0.010661   
BusinessTravel           -0.011807        1.000000  -0.015539   
DailyRate                 0.010661       -0.015539   1.000000   
DistanceFromHome         -0.001686       -0.009696  -0.004985   
Education                 0.208034       -0.008670  -0.016806   
EnvironmentSatisfaction   0.010146       -0.011310   0.018355   
HourlyRate                0.024287       -0.004164   0.023381   
JobInvolvement            0.029820        0.029300   0.046135   
JobLevel                  0.509604       -0.011696   0.002966   
JobSatisfaction          -0.004892        0.008666   0.030571   
MonthlyIncome             0.497855       -0.013450   0.007707   
MonthlyRate               0.028051       -0.008440  -0.032182   
NumCompaniesWorked        0.299635       -0.030743   0.038153   
PercentSalaryHike         0.003634       -0.025727   0.022704   
PerformanceRating         0.001904        0.001683   0.000473   
RelationshipSatisfaction  0.053535        0.008926   0.007846   
StockOptionLevel          0.037510       -0.028257   0.042143   
TotalWorkingYears         0.680381        0.007972   0.014515   
TrainingTimesLastYear    -0.019621        0.016357   0.002453   
WorkLifeBalance          -0.021490        0.004209  -0.037848   
YearsAtCompany            0.311309        0.005212  -0.034055   
YearsInCurrentRole        0.212901       -0.005336   0.009932   
YearsSinceLastPromotion   0.216513        0.005222  -0.033229   
YearsWithCurrManager      0.202089       -0.000229  -0.026363   
Churn                    -0.159205        0.127006  -0.056652   
Research & Development    0.017883        0.005308   0.014871   
Sales                    -0.027549       -0.004261  -0.003616   
Human Resources           0.001696       -0.003168  -0.043144   
Life Sciences             0.016824        0.019860   0.004028   
Marketing                 0.038162        0.005201  -0.064449   
Medical                  -0.006354       -0.011245   0.034202   
Technical Degree         -0.027604       -0.019699   0.030869   
Gender                    0.036311        0.044896   0.011716   
Married                   0.083919        0.002198   0.040035   
Single                   -0.119185        0.023017  -0.075835   
Overtime                  0.028062        0.042752   0.009135   

                          DistanceFromHome  Education  \
Age                              -0.001686   0.208034   
BusinessTravel                   -0.009696  -0.008670   
DailyRate                        -0.004985  -0.016806   
DistanceFromHome                  1.000000   0.021042   
Education                         0.021042   1.000000   
EnvironmentSatisfaction          -0.016075  -0.027128   
HourlyRate                        0.031131   0.016775   
JobInvolvement                    0.008783   0.042438   
JobLevel                          0.005303   0.101589   
JobSatisfaction                  -0.003669  -0.011296   
MonthlyIncome                    -0.017014   0.094961   
MonthlyRate                       0.027473  -0.026084   
NumCompaniesWorked               -0.029251   0.126317   
PercentSalaryHike                 0.040235  -0.011111   
PerformanceRating                 0.027110  -0.024539   
RelationshipSatisfaction          0.006557  -0.009118   
StockOptionLevel                  0.044872   0.018422   
TotalWorkingYears                 0.004628   0.148280   
TrainingTimesLastYear            -0.036942  -0.025100   
WorkLifeBalance                  -0.026556   0.009819   
YearsAtCompany                    0.009508   0.069114   
YearsInCurrentRole                0.018845   0.060236   
YearsSinceLastPromotion           0.010029   0.054254   
YearsWithCurrManager              0.014406   0.069065   
Churn                             0.077924  -0.031373   
Research & Development           -0.008117  -0.018604   
Sales                             0.014085   0.014215   
Human Resources              

MonthlyIncome and JobLevel are highly correlated (>0.9), however, monthly income is less correlated with churn so it will the one removed from the model

In [64]:
my_data = my_data.drop("MonthlyIncome", axis=1)

In [65]:
correlation_2 = my_data.corr(method = "pearson")

In [66]:
correlation_2

Age  BusinessTravel  DailyRate  \
Age                       1.000000       -0.011807   0.010661   
BusinessTravel           -0.011807        1.000000  -0.015539   
DailyRate                 0.010661       -0.015539   1.000000   
DistanceFromHome         -0.001686       -0.009696  -0.004985   
Education                 0.208034       -0.008670  -0.016806   
EnvironmentSatisfaction   0.010146       -0.011310   0.018355   
HourlyRate                0.024287       -0.004164   0.023381   
JobInvolvement            0.029820        0.029300   0.046135   
JobLevel                  0.509604       -0.011696   0.002966   
JobSatisfaction          -0.004892        0.008666   0.030571   
MonthlyRate               0.028051       -0.008440  -0.032182   
NumCompaniesWorked        0.299635       -0.030743   0.038153   
PercentSalaryHike         0.003634       -0.025727   0.022704   
PerformanceRating         0.001904        0.001683   0.000473   
RelationshipSatisfaction  0.053535        0.008926   0.007846   
StockOptionLevel          0.037510       -0.028257   0.042143   
TotalWorkingYears         0.680381        0.007972   0.014515   
TrainingTimesLastYear    -0.019621        0.016357   0.002453   
WorkLifeBalance          -0.021490        0.004209  -0.037848   
YearsAtCompany            0.311309        0.005212  -0.034055   
YearsInCurrentRole        0.212901       -0.005336   0.009932   
YearsSinceLastPromotion   0.216513        0.005222  -0.033229   
YearsWithCurrManager      0.202089       -0.000229  -0.026363   
Churn                    -0.159205        0.127006  -0.056652   
Research & Development    0.017883        0.005308   0.014871   
Sales                    -0.027549       -0.004261  -0.003616   
Human Resources           0.001696       -0.003168  -0.043144   
Life Sciences             0.016824        0.019860   0.004028   
Marketing                 0.038162        0.005201  -0.064449   
Medical                  -0.006354       -0.011245   0.034202   
Technical Degree         -0.027604       -0.019699   0.030869   
Gender                    0.036311        0.044896   0.011716   
Married                   0.083919        0.002198   0.040035   
Single                   -0.119185        0.023017  -0.075835   
Overtime                  0.028062        0.042752   0.009135   

                          DistanceFromHome  Education  \
Age                              -0.001686   0.208034   
BusinessTravel                   -0.009696  -0.008670   
DailyRate                        -0.004985  -0.016806   
DistanceFromHome                  1.000000   0.021042   
Education                         0.021042   1.000000   
EnvironmentSatisfaction          -0.016075  -0.027128   
HourlyRate                        0.031131   0.016775   
JobInvolvement                    0.008783   0.042438   
JobLevel                          0.005303   0.101589   
JobSatisfaction                  -0.003669  -0.011296   
MonthlyRate                       0.027473  -0.026084   
NumCompaniesWorked               -0.029251   0.126317   
PercentSalaryHike                 0.040235  -0.011111   
PerformanceRating                 0.027110  -0.024539   
RelationshipSatisfaction          0.006557  -0.009118   
StockOptionLevel                  0.044872   0.018422   
TotalWorkingYears                 0.004628   0.148280   
TrainingTimesLastYear            -0.036942  -0.025100   
WorkLifeBalance                  -0.026556   0.009819   
YearsAtCompany                    0.009508   0.069114   
YearsInCurrentRole                0.018845   0.060236   
YearsSinceLastPromotion           0.010029   0.054254   
YearsWithCurrManager              0.014406   0.069065   
Churn                             0.077924  -0.031373   
Research & Development           -0.008117  -0.018604   
Sales                             0.014085   0.014215   
Human Resources                  -0.002624   0.026479   
Life Sciences                    -0.024499   0.013184   
Marketing                         0.0

sales and research & developement are also highly correlated (>0.9), however, sales will be removed since it is less correlated with churn

In [67]:
my_data = my_data.drop("Sales", axis=1)

In [68]:
correlation_3 = my_data.corr(method="pearson")

In [69]:
correlation_3

Age  BusinessTravel  DailyRate  \
Age                       1.000000       -0.011807   0.010661   
BusinessTravel           -0.011807        1.000000  -0.015539   
DailyRate                 0.010661       -0.015539   1.000000   
DistanceFromHome         -0.001686       -0.009696  -0.004985   
Education                 0.208034       -0.008670  -0.016806   
EnvironmentSatisfaction   0.010146       -0.011310   0.018355   
HourlyRate                0.024287       -0.004164   0.023381   
JobInvolvement            0.029820        0.029300   0.046135   
JobLevel                  0.509604       -0.011696   0.002966   
JobSatisfaction          -0.004892        0.008666   0.030571   
MonthlyRate               0.028051       -0.008440  -0.032182   
NumCompaniesWorked        0.299635       -0.030743   0.038153   
PercentSalaryHike         0.003634       -0.025727   0.022704   
PerformanceRating         0.001904        0.001683   0.000473   
RelationshipSatisfaction  0.053535        0.008926   0.007846   
StockOptionLevel          0.037510       -0.028257   0.042143   
TotalWorkingYears         0.680381        0.007972   0.014515   
TrainingTimesLastYear    -0.019621        0.016357   0.002453   
WorkLifeBalance          -0.021490        0.004209  -0.037848   
YearsAtCompany            0.311309        0.005212  -0.034055   
YearsInCurrentRole        0.212901       -0.005336   0.009932   
YearsSinceLastPromotion   0.216513        0.005222  -0.033229   
YearsWithCurrManager      0.202089       -0.000229  -0.026363   
Churn                    -0.159205        0.127006  -0.056652   
Research & Development    0.017883        0.005308   0.014871   
Human Resources           0.001696       -0.003168  -0.043144   
Life Sciences             0.016824        0.019860   0.004028   
Marketing                 0.038162        0.005201  -0.064449   
Medical                  -0.006354       -0.011245   0.034202   
Technical Degree         -0.027604       -0.019699   0.030869   
Gender                    0.036311        0.044896   0.011716   
Married                   0.083919        0.002198   0.040035   
Single                   -0.119185        0.023017  -0.075835   
Overtime                  0.028062        0.042752   0.009135   

                          DistanceFromHome  Education  \
Age                              -0.001686   0.208034   
BusinessTravel                   -0.009696  -0.008670   
DailyRate                        -0.004985  -0.016806   
DistanceFromHome                  1.000000   0.021042   
Education                         0.021042   1.000000   
EnvironmentSatisfaction          -0.016075  -0.027128   
HourlyRate                        0.031131   0.016775   
JobInvolvement                    0.008783   0.042438   
JobLevel                          0.005303   0.101589   
JobSatisfaction                  -0.003669  -0.011296   
MonthlyRate                       0.027473  -0.026084   
NumCompaniesWorked               -0.029251   0.126317   
PercentSalaryHike                 0.040235  -0.011111   
PerformanceRating                 0.027110  -0.024539   
RelationshipSatisfaction          0.006557  -0.009118   
StockOptionLevel                  0.044872   0.018422   
TotalWorkingYears                 0.004628   0.148280   
TrainingTimesLastYear            -0.036942  -0.025100   
WorkLifeBalance                  -0.026556   0.009819   
YearsAtCompany                    0.009508   0.069114   
YearsInCurrentRole                0.018845   0.060236   
YearsSinceLastPromotion           0.010029   0.054254   
YearsWithCurrManager              0.014406   0.069065   
Churn                             0.077924  -0.031373   
Research & Development           -0.008117  -0.018604   
Human Resources                  -0.002624   0.026479   
Life Sciences                    -0.024499   0.013184   
Marketing                         0.039294   0.072405   
Medical                           0.013486  -0.072335   
Technical Degree                 -0.014802  -

performance rating and percent salary hike are highly correlated, I will remover performance rating since it is less correlated with churn

In [70]:
my_data = my_data.drop("PerformanceRating", axis=1)

In [71]:
correlation_4 = my_data.corr(method="pearson")

In [72]:
correlation_4

Age  BusinessTravel  DailyRate  \
Age                       1.000000       -0.011807   0.010661   
BusinessTravel           -0.011807        1.000000  -0.015539   
DailyRate                 0.010661       -0.015539   1.000000   
DistanceFromHome         -0.001686       -0.009696  -0.004985   
Education                 0.208034       -0.008670  -0.016806   
EnvironmentSatisfaction   0.010146       -0.011310   0.018355   
HourlyRate                0.024287       -0.004164   0.023381   
JobInvolvement            0.029820        0.029300   0.046135   
JobLevel                  0.509604       -0.011696   0.002966   
JobSatisfaction          -0.004892        0.008666   0.030571   
MonthlyRate               0.028051       -0.008440  -0.032182   
NumCompaniesWorked        0.299635       -0.030743   0.038153   
PercentSalaryHike         0.003634       -0.025727   0.022704   
RelationshipSatisfaction  0.053535        0.008926   0.007846   
StockOptionLevel          0.037510       -0.028257   0.042143   
TotalWorkingYears         0.680381        0.007972   0.014515   
TrainingTimesLastYear    -0.019621        0.016357   0.002453   
WorkLifeBalance          -0.021490        0.004209  -0.037848   
YearsAtCompany            0.311309        0.005212  -0.034055   
YearsInCurrentRole        0.212901       -0.005336   0.009932   
YearsSinceLastPromotion   0.216513        0.005222  -0.033229   
YearsWithCurrManager      0.202089       -0.000229  -0.026363   
Churn                    -0.159205        0.127006  -0.056652   
Research & Development    0.017883        0.005308   0.014871   
Human Resources           0.001696       -0.003168  -0.043144   
Life Sciences             0.016824        0.019860   0.004028   
Marketing                 0.038162        0.005201  -0.064449   
Medical                  -0.006354       -0.011245   0.034202   
Technical Degree         -0.027604       -0.019699   0.030869   
Gender                    0.036311        0.044896   0.011716   
Married                   0.083919        0.002198   0.040035   
Single                   -0.119185        0.023017  -0.075835   
Overtime                  0.028062        0.042752   0.009135   

                          DistanceFromHome  Education  \
Age                              -0.001686   0.208034   
BusinessTravel                   -0.009696  -0.008670   
DailyRate                        -0.004985  -0.016806   
DistanceFromHome                  1.000000   0.021042   
Education                         0.021042   1.000000   
EnvironmentSatisfaction          -0.016075  -0.027128   
HourlyRate                        0.031131   0.016775   
JobInvolvement                    0.008783   0.042438   
JobLevel                          0.005303   0.101589   
JobSatisfaction                  -0.003669  -0.011296   
MonthlyRate                       0.027473  -0.026084   
NumCompaniesWorked               -0.029251   0.126317   
PercentSalaryHike                 0.040235  -0.011111   
RelationshipSatisfaction          0.006557  -0.009118   
StockOptionLevel                  0.044872   0.018422   
TotalWorkingYears                 0.004628   0.148280   
TrainingTimesLastYear            -0.036942  -0.025100   
WorkLifeBalance                  -0.026556   0.009819   
YearsAtCompany                    0.009508   0.069114   
YearsInCurrentRole                0.018845   0.060236   
YearsSinceLastPromotion           0.010029   0.054254   
YearsWithCurrManager              0.014406   0.069065   
Churn                             0.077924  -0.031373   
Research & Development           -0.008117  -0.018604   
Human Resources                  -0.002624   0.026479   
Life Sciences                    -0.024499   0.013184   
Marketing                         0.039294   0.072405   
Medical                           0.013486  -0.072335   
Technical Degree                 -0.014802  -0.026742   
Gender                            0.001851   0.016547   
Married                           0.030232  -0.001865

there are some variables left that are also correlated, but they have relatively high correlations with churn so i's rather not remove them

now let's try to create the algorithms. 

In [74]:
ibm_logit = LogisticRegression(random_state=42)
ibm_tree = DecisionTreeClassifier(random_state=42)

In [75]:
y = my_data.Churn #what we want to predict
x= my_data.drop("Churn", axis=1)

In [76]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.25,random_state=42)

In [78]:
ibm_logit.fit(x_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=42, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [79]:
ibm_logit.score(x_train,y_train)*100

88.56624319419237

In [80]:
ibm_logit.score(x_test,y_test)*100

89.94565217391305

low accuracy, yet no (or maybe not a lot because idk what it means when overfitting is higher than accuracy) overfitting. I will try to get higher accuracy though. 

In [82]:
ibm_tree.fit(x_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=42,
            splitter='best')

In [83]:
ibm_tree.score(x_train,y_train)*100


100.0

In [84]:
ibm_tree.score(x_test,y_test)*100 

76.63043478260869

as we can see, there is high overfitting in this case. 

we can use some tools to fight this overfitting problem such as: limiting the tree's growth, implementing a minimum sample size and so on. we should also see if we have a big disbalance between churners and non-churners (probably we do) and balance them to get a higher recall score. 

In [85]:
100.0*my_data.Churn.value_counts()/len(my_data)

0    83.877551
1    16.122449
Name: Churn, dtype: float64

that classifies as big disbalance, so I will keep that in mind

In [86]:
ibm_tree_7 = DecisionTreeClassifier(max_depth = 7, random_state=42)
ibm_tree_7.fit(x_train, y_train)
ibm_tree_7.score(x_test, y_test)*100

83.15217391304348

In [88]:
ibm_tree_7.score(x_train, y_train)*100 #forgot this in the beginning

93.92014519056261

less accurate, still there is overfitting (not as bad as the first one)

In [89]:
ibm_tree_8 = DecisionTreeClassifier(max_depth = 8, random_state=42)
ibm_tree_8.fit(x_train, y_train)
ibm_tree_8.score(x_train, y_train)*100

95.73502722323049

In [136]:
ibm_tree_8.score(x_test, y_test)*100#more accurate yet more overfitting

82.6086956521739

In [181]:
ibm_tree_leaf = DecisionTreeClassifier(min_samples_leaf=40, max_depth = 8, random_state=42)
ibm_tree_leaf.fit(x_train, y_train)
ibm_tree_leaf.score(x_train, y_train)*100

85.39019963702358

In [182]:
ibm_tree_leaf.score(x_test, y_test)*100

85.32608695652173

tested many variations of the last one (ibm_tree_leaf), that's the best one I could come up with

In [102]:
ibm_tree_8_b = DecisionTreeClassifier(class_weight="balanced", max_depth=8, random_state=42)
ibm_tree_8_b.fit(x_train, y_train)
ibm_tree_8_b.score(x_train, y_train)*100


90.65335753176043

In [103]:
ibm_tree_8_b.score(x_test, y_test)

0.7527173913043478

pretttttty bad

In [104]:
from sklearn.model_selection import cross_val_score


In [105]:
np.mean(cross_val_score(ibm_tree_8, x, y, cv=10))


0.8163585797147441

In [106]:
np.mean(cross_val_score(ibm_tree_7, x, y, cv=10))

0.8224671764299945

In [107]:
from sklearn.metrics import recall_score


In [108]:
prediction_7 = ibm_tree_7.predict(x_test)

In [110]:
recall_score(y_test,prediction_7)*100

14.583333333333334

In [112]:
prediction_8 = ibm_tree_8.predict(x_test)
recall_score(y_test, prediction_8)*100

18.75

In [113]:
prediction_logit = ibm_logit.predict(x_test)
recall_score(y_test, prediction_logit)*100

39.58333333333333

much better recall score, but i'm still not satisfied with this result. I will try to implement a loop to find which parameters will give me the best model in case of decision trees, maybe I will find better models (basically what we learned in class today). 

In [114]:
from sklearn.model_selection import GridSearchCV

In [115]:
ibm_test = DecisionTreeClassifier(class_weight="balanced", random_state=42)

In [116]:
max_depth_values = [i for i in range(5,21)]
min_sample_values = [i for i in range(10,501,10)]

In [118]:
parameters={"max_depth":max_depth_values,"min_samples_leaf": min_sample_values}

In [119]:
finder = GridSearchCV(ibm_test, parameters)

In [120]:
finder.fit(x_train, y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight='balanced', criterion='gini',
            max_depth=None, max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=42,
            splitter='best'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'max_depth': [5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20], 'min_samples_leaf': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200, 210, 220, 230, 240, 250, 260, 270, 280, 290, 300, 310, 320, 330, 340, 350, 360, 370, 380, 390, 400, 410, 420, 430, 440, 450, 460, 470, 480, 490, 500]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [121]:
print(finder.best_params_)

{'max_depth': 5, 'min_samples_leaf': 370}


In [132]:
ibm_tree_best = DecisionTreeClassifier(max_depth = 5, min_samples_leaf=370, class_weight = "balanced", random_state=42)
ibm_tree_best.fit(x_train, y_train)
ibm_tree_best.score(x_train, y_train)*100

65.60798548094374

In [133]:
ibm_tree_best.score(x_test, y_test)*100

68.75

basically this model is not good at all because the accuracy is very low. I tried other parameters, still nothing works.

In [134]:
from sklearn.metrics import roc_auc_score
roc_auc_score(y_test,prediction_logit)*100

68.54166666666667

In [135]:
roc_auc_score(y_test, prediction_8)*100

55.46875

In [183]:
prediction_leaf = ibm_tree_leaf.predict(x_test)

In [184]:
recall_score(y_test, prediction_leaf)*100

10.416666666666668

In [186]:
roc_auc_score(y_test, prediction_leaf)*100

53.489583333333336

Conclusion from all this: the logistic regression model serves as the best algorithm that i could find to predict churn; it has no overfitting problems, higher recall and roc_auc scores than the other models and an acceptable accuracy. 